### ISSUE TITLE
Cannot SSH to container to MC application

### ISSUE THREAD

**Issue by ionita-sonah** [balena-io/balena-cli#856]
Hey, I am trying to connect to a container, part of a multi-container application. This works via the dashboard terminal, but I don't like using the browser long-term to check stuff and would prefer the ubuntu terminal.

I first tried using resin-cli (version: 7.0.7):
`resin ssh <device_uuid>`
I get "ResinRequestError: Request error: Some v1 endpoints are only allowed on single-container apps"

Then I tried connecting to the host OS (using the `--host` option), which succeded. Then used balena to list the running containers (`balena ps`), then attach to my preferred one (`balena attacth <container_id>`). The command hangs however and I never get attached. 

Any ideas how I can connect? 

<img src="https://frontapp.com/assets/img/icons/favicon-32x32.png" height="16" width="16" alt="Front logo" /> [Front conversations](https://app.frontapp.com/open/top_e3ej)

**Comment by pimterry** [comment]
Hi @ionita-sonah, you're totally right. The CLI doesn't yet fully support all functionality on multicontainer devices, but we're working on it at the moment.

As a workaround though, you can indeed SSH through the host. The command you're looking for after SSHing is:

```bash
balena exec -it <container id> /bin/sh
```

passing the container id from `balena ps`.

**Comment by ionita-sonah** [comment]
Hi @pimterry, this was exactly what I was looking for. Cheers!

**Comment by pimterry** [comment]
Ah, just spotted that this shouldn't have been closed. There's a workaround above, but the issue itself does still need fixing.

**Comment by memory** [comment]
Thank you, we're definitely interesting in seeing this fixed!

**Comment by lasote** [comment]
I can't run `balena ps` nor `balena exec`. Have these commands been removed or am I doing wrong?

**Comment by pimterry** [comment]
@lasote I think you've confused balena cli with balena engine (https://github.com/balena-os/balena-engine). We recently [renamed the company](https://www.balena.io/blog/resin-io-changes-name-to-balena-releases-open-source-edition/) from resin.io to balena, and the names of a few components have changed en route.

Balena cli (previously resin cli) lets you manage and provision balena devices. Balena engine (previously just balena) runs on the devices themselves, and is a moby-based container engine similar to Docker. Balena cli has never supported the commands you're looking for, but the engine running on your device does.

Balena cli used to be run as `resin`, but is now `balena`. Balena engine used to be run as `balena`, but is now `balena-engine`. You're probably looking for `balena-engine ps` and `balena-engine exec`.

**Comment by KingRial** [comment]
Is there any way from open-balena service to enable/disable an SSH service on a specific device based on "balena OS" production image ?

If so, while waiting for an official support using balena-cli, any remote support could be handled by using the open-balena-vpn container and launching SSH over the desired VPN IP and the desired port.

**Comment by richbayliss** [comment]
FYI @hedss, this issue is known already

**Comment by pdcastro** [comment]
This feature was implemented in CLI v11 (currently v11.18.2).
```
$ balena help ssh
Usage: ssh <applicationOrDevice> [serviceName]

This command can be used to start a shell on a local or remote device.

If a service name is not provided, a shell will be opened on the host OS.

If an application name is provided, all online devices in the application
will be presented, and the chosen device will then have a shell opened on
in it's service container or host OS.

For local devices, the IP address and .local domain name are supported.
If the device is referenced by IP or `.local` address, the connection
is initiated directly to balenaOS on port `22222` via an
openssh-compatible client. Otherwise, any connection initiated remotely
traverses the balenaCloud VPN.

Examples:
	balena ssh MyApp

	balena ssh f49cefd
	balena ssh f49cefd my-service
	balena ssh f49cefd --port <port>

	balena ssh 192.168.0.1 --verbose
	balena ssh f49cefd.local my-service

Warning: `balena ssh` requires an openssh-compatible client to be correctly
installed in your shell environment. For more information (including Windows
support) please check:
	https://github.com/balena-io/balena-cli/blob/master/INSTALL.md#additional-dependencies

Options:

    --noproxy                           Don't use the proxy configuration for this connection. This flag only make sense if you've configured a proxy globally.
    --port, -p <port>                   SSH gateway port
    --verbose, -v                       Increase verbosity
```
